### ISSUE TITLE
allow partial allocation success

### ISSUE THREAD

**Issue by ahelsing** [GENI-NSF/geni-tools#563]
If your request has > 1 circuits, you might want stitcher to complete the allocation of some of the circuits, even if 1 fails. Can we provide a way to do this?

stitcher would have to ensure that all AMs that are required for the successful VLANs are up. And what about the resources in some of those AMs that are for the otherwise failed circuit? And what happens to the manifest?

 _Imported from trac ticket [#563](http://trac.gpolab.bbn.com/gcf/ticket/563),  created by ahelsing on 03-18-2014 at 23:57, last modified: 05-12-2015 at 15:54_

**Comment by ahelsing** [comment]
How common is it that a request has multiple non overlapping circuits? I assume not very.

So the issue is about helping a circuit come up the rest of the way.

Is this useful if a particular circuit only comes up part way? What would you do in this case? Retry manually the parts that failed? That's only useful if you assume that a retry will fix things (say, by waiting a bit before doing that retry), or that a human could edit the request so it will succeed without breaking the rest of the circuit. (Say, a typo in the AM specific nodes.)

What would have to happen?
 - when you get a failure, do not delete all reservations that succeeded. Delete only the failed bits
 - Give a manifest of what you did get - but ideally only what you actually got
 - Give the proper request RSpec for the failed AM(s). Ideally a single request that can be edited to work for all the AMs that have not been reserved yet, modulo editing VLAN tags
 - What does an experimenter do with that though? Run it manually through Omni? That would be a good first step. Even better, stitcher can somehow take as input a request/manifest hybrid plus the previous SCS workflow that it can use to tell which AMs still require a reservation and pick up where it left off.


_Trac comment by ahelsing on 10-17-2014 at 14:10_

**Comment by ahelsing** [comment]
This issue and desire came up again during GEC22.

Note that if most links cross AL2S, this could simply be a mode where stitcher does not make the reservations at AL2S (or perhaps any transit network).

If this is a single AM, then presumably there's a single request that has not been completed. So a single manifest for everything else plus a single request.

This has the benefit that you could glue the pieces together with multiple slices if desired.


_Trac comment by ahelsing on 03-30-2015 at 10:53_

**Comment by ahelsing** [comment]
Ezra says that the simple 'no AL2S' option is what he does manually now. So I guess that's of some value. (And easier when there is stitching to LANs.) But for hime the issue is not having to wait to delete resources across multiple AMs if somewhere in the list the BW is unavailable or AL2S forces a new tag to be picked or something. Only putting off AL2S isn't enough; the AM that cannot give the BW is also an issue.

So I think Ezra wants stitcher to optionally not delete everything on partial failure. The question though is what it gives you in this case in terms of support for fixing & building what remains.

For example, when stitcher is otherwise about to delete, it could stop, and output
 - combined manifest so far
 - list of AMs where you have reservations
 - list of AMs where it failed
 - list of AMs where it has not attempted
 - for each failure, list of VLANs unavailable and/or error message that was fatal
 - for each AM where no reservation, a template request (filling in all the VLAN tags that we do know so far)

And then what? The experimenter can do the rest manually. But can stitcher help?
 - Some mode where it takes a request, hands the SCS the request with the proper VLAN unavailable tags, is initialized to know which AMs you already have a reservation at, and only attempts reservations where necessary. This is sort of a "resume" function. Takes an expanded request, contacts the AMs to see which already have reservations in the slice like for recreating the combined manifest, and then continues as though this is a new reservation.
 - A simple mode that only talks to a single given AM, retrying differrent VLAN tags if necessary.

_Trac comment by ahelsing on 04-08-2015 at 16:29_

**Comment by ahelsing** [comment]
see ticket #791

_Trac comment by ahelsing on 04-22-2015 at 14:03_

**Comment by ahelsing** [comment]
Ezra confirms that his current workflow reserves nodes separately from stitched links essentially. So a stitcher mode that split these up somehow would be useful.

A big win would be an option that does not "delete everything" on failure. Specifically, his issue is when one AM cannot give the BW he wants.

So:
 - An option to not delete everything on failure but instead report what you have and stop
 - Check avail BW where possible and complain if it isn't possible before starting
 - Options to allow requesting a range of BW, and when you get the insufficient BW error, auto retry something smaller within range allowed by experimenter (decrements specified by user)
 - More generally, some kind of rspec extension allowing you to specify a range for various things (set of possible component IDs, etc)

But Ezra seems to be saying to not stress about stitcher auto picking up from where you left off.

So as a first pass:
If you get some new option, and a failure is going to be fatal:
 - can stitcher categorize the range of fatal errors between those that are user correctable and those that are not?
 - Otherwise, on fatal error or Ctrl-C, do not delete all.
 - Instead
  - write a manifest. This manifest starts from some AM that has a manifest, but merges in also all requests. In the comments, list the AMs where there are reservations and those where there are not. Also list the AM with the fatal error and what it was.
   - Ideally, comments should also indicate which AMs depend on which other AMs, and/or which hops depend on which other hops
  - write a request. This is identical to the manifest, but starts from the AM that failed, and so is essentially a request.

Could have another option: NoTransitAggregates. With this option, stitcher stops when the last few AMs are all SCS added Then it acts like the above Ctrl-C case, except there is no error to report.


_Trac comment by ahelsing on 04-22-2015 at 14:15_

**Comment by ahelsing** [comment]
Adding 2 options: `--noDeleteAtEnd` (see ticket #791) and `--noTransitAMs`.

`--noTransitAMs` checks if the only remaining ready (dependencies satified) AMs are not user requested (so typically transit AMs like AL2S). If so, log something and raise an error. That kicks the `--noDeleteAtEnd` functionality.

`--noDeleteAtEnd`: when true, instead of delete all reservations:
 - log something about how many AMs have reservations, how many do not, how many failed
 - write a combined manifest RSpec for what we do have
 - write a combined request RSpec for what we do not have
 - both those RSpecs include things like the VLANs unavailable, where a hop should take its VLAN tag from, what the last error was from an AM if any.

An experimenter would have to edit that request RSpec to fill in the proper new VLAN tags to complete the reservation.

_Trac comment by ahelsing on 05-12-2015 at 14:58_

**Comment by ahelsing** [comment]
A separate ticket tracks checking available bandwidth: #809

The idea of a stitcher resume function is in ticket #810


_Trac comment by ahelsing on 05-12-2015 at 15:54_
